tools/generator/src/Extensions/XcodeScheme+Extensions.swift

```diff
@@ -25,154 +25,159 @@ extension XcodeScheme {
         }
 
         return resolvedTargetIDs
     }
 }
 
-// MARK: LabelValueTargetInfo
+extension TargetResolver {
+    var labelTargetInfos: [BazelLabel: XcodeScheme.LabelTargetInfo] {
+        get throws {
+            var results = [BazelLabel: XcodeScheme.LabelTargetInfo]()
+
+            // Collect the target information
+            for (targetID, target) in targets {
+                let targetWithID = XcodeScheme.TargetWithID(id: targetID, target: target)
+                var targetInfo = try results[target.label] ?? .init(
+                    label: target.label,
+                    isTopLevel: try pbxTargetAndKey(for: targetID).pbxTarget.isTopLevel
+                )
+                targetInfo.add(targetWithID)
+                results[target.label] = targetInfo
+            }
+
+            return results
+        }
+    }
+}
+
+// MARK: LabelTargetInfo
 
 extension XcodeScheme {
     /// Collects Target information for a BazelLabel.
-    struct LabelValueTargetInfo {
+    struct LabelTargetInfo {
         let label: BazelLabel
+        let isTopLevel: Bool
         var inPlatformOrder = [TargetWithID]()
         var platforms = Set<Platform>()
+    }
+}
 
-        func best() throws -> TargetWithID {
-            guard let best = inPlatformOrder.first else {
-                throw PreconditionError(message: """
+extension XcodeScheme.LabelTargetInfo {
+    mutating func add(_ targetWithID: XcodeScheme.TargetWithID) {
+        inPlatformOrder.append(targetWithID)
+        inPlatformOrder.sort()
+        platforms.update(with: targetWithID.target.platform)
+    }
+}
+
+extension XcodeScheme.LabelTargetInfo {
+    var best: XcodeScheme.TargetWithID {
+        get throws {
+            return try inPlatformOrder.first.orThrow("""
 Unable to find the best `TargetWithID` for "\(label)"
 """)
-            }
-            return best
         }
     }
+}
 
-    private func collectTargetInfoByLabelValue(
-        targetWithIDs: [TargetWithID]
-    ) -> [BazelLabel: LabelValueTargetInfo] {
-        var results = [BazelLabel: LabelValueTargetInfo]()
-
-        // Collect the target information
-        for targetWithID in targetWithIDs {
-            let target = targetWithID.target
-            var targetInfo = results[
-                target.label, default: LabelValueTargetInfo(label: target.label)
-            ]
-            targetInfo.platforms.update(with: target.platform)
-            targetInfo.inPlatformOrder.append(targetWithID)
-            targetInfo.inPlatformOrder.sort()
-            results[target.label] = targetInfo
-        }
-
-        return results
+extension XcodeScheme.LabelTargetInfo {
+    func firstCompatibleWith<Platforms: Sequence>(
+        anyOf platforms: Platforms
+    ) -> XcodeScheme.TargetWithID? where Platforms.Element == Platform {
+        let uniquePlatforms = Set(platforms)
+        return inPlatformOrder.first { uniquePlatforms.contains($0.target.platform) }
```

Comment:
Just confirming, does this pick the "best" from the topLevelPlatforms? Say we have tvOS and iOS top level targets and a library that is a dependency of both, will this always pick the iOS version of the library?

Reply:
I now see the `inPlatformOrder` 👍.

Reply:
Yes. I added a comment where we pass in the top-level platforms. The `inPlatformOrder` array maintains the list of targets in `Platform` order. Hence, for your example, the first one that we find should be iOS. Also, #856 introduces a better compatibility check to evaluate the minimum OS version as a less-than-or-equal-to.